torch/ao/sparsity/sparsifier.py

```diff
@@ -0,0 +1,167 @@
+        new_name = prefix + '.' + name
+        if child is layer:
+            return new_name
+        child_path = _module_to_path(child, layer, prefix=new_name)
+        if child_path is not None:
+            return child_path
+    return None
+
+def _path_to_module(model, path):
+    path = path.split('.')
+    for name in path:
+        model = getattr(model, name, None)
+        if model is None:
+            return None
+    return model
+
+
+class BaseSparsifier(abc.ABC):
+    r"""Base class for all sparsifiers.
+
+    Abstract methods that need to be implemented:
+
+    - step: Function to compute a new mask for all keys in the `module_groups`.
+
+    Args:
+        - model [nn.Module]: model to configure. The model itself is not saved
+            but used for the state_dict saving / loading.
+        - config [list]: configuration elements could either be instances of
+            nn.Module or dict maps. The dicts must have a key 'module' with the
+            value being an instance of a nn.Module.
+        - defaults [dict]: default configurations will be attached to the
+            configuration. Only the keys that don't exist in the `config` will
+            be updated.
+
+    Example::
+
+        >>> config = [model.layer1, {'module': model.linear2, 'sparsity_level': 0.5}]
+        >>> defaults = {'sparsity_level': 0.7}
+        >>> # model.layer1 will have `sparsity_level` = 0.7 (getting default)
+        >>> sparsifier = BaseSparsifier(config, defaults)
+    """
+    def __init__(self, model, config, defaults):
+        super().__init__()
+        self.config = config
+        self.defaults = defaults
+        if self.defaults is None:
+            self.defaults = dict()
+
+        self.state = {}
+        self.module_groups = []
+        self.enable_mask_update = False
+
+        self.model = model
+        # If no config -- try getting all the supported layers
+        if self.config is None:
+            # Add all models to the config
+            self.config = []
+            stack = [model]
+            while stack:
+                module = stack.pop()
+                for name, child in module.named_children():
+                    if type(child) in SUPPORTED_MODULES:
+                        self.config.append(child)
+                    else:
+                        stack.append(child)
+
+        for module_config in self.config:
+            if isinstance(module_config, nn.Module):
+                module_config = {'module': module_config}
+            local_args = copy.deepcopy(self.defaults)
+            local_args.update(module_config)
+            module = local_args['module']
+            module_path = _module_to_path(self.model, module)
+            if module_path and module_path[0] == '.':
+                module_path = module_path[1:]
+            local_args['path'] = module_path
+            self.module_groups.append(local_args)
+
+    def __getstate__(self):
```

Comment:
Do we intend for sparsifier objects to be pickleable? module groups contain references to modules, so they shouldnt be pickleable

Reply:
At this stage the module_groups have only the FQNs, which we want to pickle, right?

Reply:
Yes, that is correct.